File: integrations/examples/python/mcp_boundary.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any

from intel.integration import hash_identifier
from intel.middleware import MCPToolGuard, VexylRequestGuard

DOCS_SERVER_NAME = "docs"
DOCS_TOOL_NAME = "search"
DOCS_TOOL_ACTION = "search approved documentation"


def create_docs_search_guard(request_guard: VexylRequestGuard) -> MCPToolGuard:
    """Create an MCP guard from static application-owned authorization policy."""

    return MCPToolGuard(
        request_guard,
        server_name=DOCS_SERVER_NAME,
        tool_name=DOCS_TOOL_NAME,
        tool_action=DOCS_TOOL_ACTION,
        permissions=["read"],
        user_allowed_actions=[DOCS_TOOL_ACTION],
        policy_allowed_actions=[DOCS_TOOL_ACTION],
        verified_mitigations=[
            "tool_allowlist",
            "scoped_read_only_credentials",
        ],
    )


async def execute_docs_search(
    request_guard: VexylRequestGuard,
    *,
    query: str,
    session_reference: str,
    identifier_key: str | bytes,
    execute: Callable[[str], Awaitable[Any]],
) -> Any:
    """Authorize a fixed read-only MCP action before invoking its handler."""

    validated_query = _validated_query(query)
    session_hash = hash_identifier(
        _bounded_reference(session_reference), identifier_key
    )
    await create_docs_search_guard(request_guard).authorize(
        "Search the approved documentation corpus with a validated query.",
        session_id_hash=session_hash,
        data_origin="internal_db",
        data_classification="internal",
    )
    return await execute(validated_query)


def _validated_query(value: str) -> str:
    if not isinstance(value, str):
        raise TypeError("query must be a string")
    if any(ord(character) < 32 for character in value):
        raise ValueError("query contains control characters")
    normalized = " ".join(value.split())
    if not normalized or len(normalized) > 200:
        raise ValueError("query must be between 1 and 200 characters")
    return normalized


def _bounded_reference(value: str) -> str:
    if not isinstance(value, str):
        raise TypeError("session_reference must be a string")
    if not value.strip() or len(value) > 512:
        raise ValueError("session_reference must be a bounded non-empty string")
    return value

File: integrations/examples/node/mcp-boundary.mjs
import { hashIdentifier } from "../../node/vexyl-guard-client.mjs";
import { MCPToolGuard } from "../../node/vexyl-guard-middleware.mjs";

export const DOCS_SERVER_NAME = "docs";
export const DOCS_TOOL_NAME = "search";
export const DOCS_TOOL_ACTION = "search approved documentation";

export function createDocsSearchGuard(requestGuard) {
  return new MCPToolGuard(requestGuard, {
    serverName: DOCS_SERVER_NAME,
    toolName: DOCS_TOOL_NAME,
    toolAction: DOCS_TOOL_ACTION,
    permissions: ["read"],
    userAllowedActions: [DOCS_TOOL_ACTION],
    policyAllowedActions: [DOCS_TOOL_ACTION],
    verifiedMitigations: ["tool_allowlist", "scoped_read_only_credentials"],
  });
}

export async function executeDocsSearch(
  requestGuard,
  { query, sessionReference, identifierKey, execute },
) {
  const validatedQuery = validateQuery(query);
  const sessionHash = hashIdentifier(
    boundedReference(sessionReference),
    identifierKey,
  );
  await createDocsSearchGuard(requestGuard).authorize(
    "Search the approved documentation corpus with a validated query.",
    {
      sessionIdHash: sessionHash,
      dataOrigin: "internal_db",
      dataClassification: "internal",
    },
  );
  if (typeof execute !== "function") throw new TypeError("execute must be a function");
  return await execute(validatedQuery);
}

function validateQuery(value) {
  if (typeof value !== "string") throw new TypeError("query must be a string");
  if (/[\u0000-\u001f]/.test(value)) {
    throw new TypeError("query must not contain control characters");
  }
  const normalized = value.trim().split(/\s+/).join(" ");
  if (!normalized || normalized.length > 200) {
    throw new TypeError("query must be a bounded non-empty string");
  }
  return normalized;
}

function boundedReference(value) {
  if (typeof value !== "string" || !value.trim() || value.length > 512) {
    throw new TypeError("sessionReference must be a bounded non-empty string");
  }
  return value;
}
